src/ddmd/statementsem.d

```diff
@@ -2490,6 +2489,19 @@ else
                         cs.error("case variables not allowed in final switch statements");
                         errors = true;
                     }
+
+                    /* Also check if the VarExp is declared in a scope outside of this one
+                     * 'sc2' is set to the scope of the switch statement.
+                     */
+                    Scope* sc2 = sc;
+                    while (sc2.enclosing && sc2.enclosing.sw == sw)
```

Comment:
I'd rather this was turned into a `for` loop where the scope of `sc2` does not extend beyond the `for` statement's scope.